### DISCUSSION TITLE
Implementation of the family of QuAPI methods

### DISCUSSION THREAD

**Discussion by amartyabose** [qutip/qutip#1718] [Ideas]
Hi,

Would there be interest in adding the quasi-adiabatic propagator path integral (QuAPI) family of methods for quantum dynamics under the system-solvent decomposition to QuTiP?  In addition to the usual algorithms, there are comparatively new tensor network-based algorithms that can also be incorporated using TensorFlow in the long term.  Of course, I understand that it would take a while to get all of this implemented, but I just wanted to check for general interest.

Best,
Amartya

**Comment by hodgestar** [comment]
@amartyabose This sounds great. A few thoughts:

* We should target QuTiP v5 for this (i.e. PRs should be against dev.major).
* There is work starting on a tensor network backend for QuTiP which might be helpful later (see https://github.com/qutip/qutip-tensornetwork).
* I guess it would live under `qutip.nonmarkov.quapi`.

Happy to have a call and potentially to help with the coding. I'm fairly busy with HEOM and QuTiP v5 alpha until the end of the year, so while I'm happy to chat before that, I'd likely only have significant time to contribute after holidays end.

**Comment by amartyabose** [comment]
@hodgestar 

I can completely understand.  Adding HEOM would be a full time work.  Are you planning on adding some of the more recent developments that incorporate spectral densities apart from a Drude?  I would love to talk with you one of these days.  Let's try to fix some time.

Till then, would you be able to guide me where I can start to familiarize myself with the basics of QuTiP?  I have always written my own quantum dynamics codes from scratch and never really tinkered around with QuTiP.  If there are small things that I can do to help out and simultaneously get to know the codebase, please let me know.

Best.

**Comment by hodgestar** [comment]
@amartyabose Let's try arrange a quick chat for this coming week. You're welcome to email me at my GitHub nickname at gmail.

The HEOM additions I'm working on currently bring the solver described in https://arxiv.org/abs/2010.10806 (preprint) into QuTiP (aka "the BoFiN solver"). The mathematics is the same as in the paper, but I rewrote the code substantially when incorporating it into QuTiP.

The new code supports fairly arbitrary bosonic and fermionic baths via an exponential expansion of the bath correlation functions. There are helper classes for constructing some common baths (https://github.com/qutip/qutip/blob/master/qutip/nonmarkov/bofin_baths.py) but one can also specify the expansion coefficients directly if needed.

The implementation itself is fairly stable now, but I'm still working on the documentation for it in https://github.com/qutip/qutip/pull/1724.

Let's chat a bit about how to get you involved with QuTiP during the call, but one good place to start is just to play around with some of the tutorials that seem interesting to you -- https://qutip.org/tutorials.html. The tutorials are not always the most up to date, but the core concepts are all still relevant and it's a fine way to start.

**Comment by Jappenn** [comment]
Could be of interest the simulation of axion-photon conversion,with the target to know the origins of cosmic axion.

**Reply by Jappenn** [reply]
Exist,now in the world,an experiment on searching dark matter.
